openid/openid.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENSE file for details.
+
+// Package openid contains functions to help log-in to Ubuntu SSO using
+// OpenID 2.0.
+package openid
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/yohcop/openid-go"
+
+	"github.com/juju/usso"
+)
+
+const (
+	// These standard simple registration fields are supported by
+	// Ubuntu SSO.
+	SimpleRegistrationNickname = "nickname"
+	SimpleRegistrationEmail    = "email"
+	SimpleRegistrationFullName = "fullname"
+	SimpleRegistrationPostcode = "postcode"
+	SimpleRegistrationCountry  = "country"
+	SimpleRegistrationLanguage = "language"
+	SimpleRegistrationTimezone = "timezone"
+
+	// These non-standard simple registration fields are supported by
+	// Ubuntu SSO.
+	SimpleRegistrationAddress1 = "x_address1"
+	SimpleRegistrationAddress2 = "x_address2"
+	SimpleRegistrationCity     = "x_city"
+	SimpleRegistrationProvince = "x_province"
+	SimpleRegistrationPhone    = "x_phone"
+)
+
+const (
+	nsSimpleRegistration = "http://openid.net/extensions/sreg/1.1"
+	nsTeams              = "http://ns.launchpad.net/2007/openid-teams"
+)
+
+// NonceStore is the NonceStore type from github.com/yohcop/openid-go. It
+// is replicated here for the convenience of clients.
+type NonceStore interface {
+	openid.NonceStore
+}
+
+// DefaultNonceStore contains the default NonceStore to use when
+// verifying OpenID responses from Ubuntu SSO if nil is passed as the ns
+// parameter of Verify. The initial value of DefaultNonceStore is an
+// in-memory store.
+var DefaultNonceStore NonceStore = openid.NewSimpleNonceStore()
+
+// DiscoveryCache is the DiscoveryCache type from
+// github.com/yohcop/openid-go. It is replicated here for the convenience
+// of clients.
+type DiscoveryCache interface {
+	openid.DiscoveryCache
+}
+
+// DefaultDiscoveryCache contains the openid.DiscoveryCache to use when
+// verifying OpenID responses from Ubuntu SSO. If this is not set then
+// Verify will use a cache that generates the discovery information in
+// the same way as Ubuntu SSO.
+var DefaultDiscoveryCache DiscoveryCache
```

Comment:
I'm not sure about having this as a global default. It's kind-of OK when there's a useful non-nil default (as there is for DefaultNonceStore, although I'm not that sure about that either), but having this as a default encourages code to put runtime state in global variables, which I'm pretty sure is a bad idea in general.